src/main/java/com/fcgl/madrid/points/service/UserActionService.java

```diff
@@ -1,22 +1,96 @@
 package com.fcgl.madrid.points.service;
 
-import com.fcgl.madrid.points.model.UserAction;
-import com.fcgl.madrid.points.repository.UserActionRepository;
+import com.fcgl.madrid.points.model.*;
+import com.fcgl.madrid.points.repository.*;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
 import org.springframework.stereotype.Service;
 
 import java.util.List;
 
 @Service
 public class UserActionService {
     private UserActionRepository userActionRepository;
+    private ActionRepository actionRepository;
+    private TrophyRepository trophyRepository;
+    private UserTrophyRepository userTrophyRepository;
+    private UserPointsRepository userPointsRepository;
 
     @Autowired
-    public UserActionService(UserActionRepository userActionRepository) {
+    public UserActionService(UserActionRepository userActionRepository,
+                             ActionRepository actionRepository,
+                             TrophyRepository trophyRepository,
+                             UserTrophyRepository userTrophyRepository,
+                             UserPointsRepository userPointsRepository) {
         this.userActionRepository = userActionRepository;
+        this.actionRepository = actionRepository;
+        this.trophyRepository = trophyRepository;
+        this.userTrophyRepository = userTrophyRepository;
+        this.userPointsRepository = userPointsRepository;
     }
 
     public List<UserAction> findAll() {
         return userActionRepository.findAll();
     }
+
+    public ResponseEntity<Trophy> addUserAction(Long userID, Long actionID) {
+
+        // Check that the action exists
+        Action action = actionRepository.findById(actionID).orElse(null);
+        if (action == null) {
+            return ResponseEntity.badRequest().body(null);
+        }
+
+        UserAction userAction = userActionRepository.findByUserIdAndAction(userID, action).orElse(null);
```

Comment:
I think we could divide this function into smaller private functions? What do you think?

Line 44 - Line 54 could be in a function that returns a `UserAction`. Maybe: `getUserAction(userId, action)`

Reply:
done!